contracts/Aavegotchi/facets/ERC721BuyOrderFacet.sol

```diff
@@ -0,0 +1,191 @@
+        .erc721TokenToBuyOrderIds[_erc721TokenAddress][_erc721TokenId].length;
+        s.erc721TokenToBuyOrderIds[_erc721TokenAddress][_erc721TokenId].push(buyOrderId);
+        s.buyerToBuyOrderId[_erc721TokenAddress][_erc721TokenId][sender] = buyOrderId;
+
+        s.erc721BuyOrders[buyOrderId] = ERC721BuyOrder({
+            buyOrderId: buyOrderId,
+            buyer: sender,
+            erc721TokenAddress: _erc721TokenAddress,
+            erc721TokenId: _erc721TokenId,
+            priceInWei: _priceInWei,
+            validationHash: LibBuyOrder.generateValidationHash(_erc721TokenAddress, _erc721TokenId, _validationOptions),
+            timeCreated: block.timestamp,
+            timePurchased: 0,
+            cancelled: false,
+            validationOptions: _validationOptions
+        });
+        emit ERC721BuyOrderAdd(buyOrderId, sender, _erc721TokenAddress, _erc721TokenId, category, _priceInWei, block.timestamp);
+    }
+
+    function cancelERC721BuyOrder(uint256 _buyOrderId) external {
+        address sender = LibMeta.msgSender();
+        ERC721BuyOrder memory erc721BuyOrder = s.erc721BuyOrders[_buyOrderId];
+
+        require(erc721BuyOrder.timeCreated != 0, "ERC721BuyOrder: ERC721 buyOrder does not exist");
+        require(
+            (sender == s.aavegotchis[erc721BuyOrder.erc721TokenId].owner) || (sender == erc721BuyOrder.buyer),
+            "ERC721BuyOrder: Only aavegotchi owner or buyer can call this function"
+        );
+        require((erc721BuyOrder.cancelled == false) && (erc721BuyOrder.timePurchased == 0), "ERC721BuyOrder: Already processed");
+
+        LibBuyOrder.cancelERC721BuyOrder(_buyOrderId);
+    }
+
+    function executeERC721BuyOrder(uint256 _buyOrderId) external {
+        address sender = LibMeta.msgSender();
+        ERC721BuyOrder memory erc721BuyOrder = s.erc721BuyOrders[_buyOrderId];
+
+        require(erc721BuyOrder.timeCreated != 0, "ERC721BuyOrder: ERC721 buyOrder does not exist");
+        require(sender == s.aavegotchis[erc721BuyOrder.erc721TokenId].owner, "ERC721BuyOrder: Only aavegotchi owner can call this function");
+        require((erc721BuyOrder.cancelled == false) && (erc721BuyOrder.timePurchased == 0), "ERC721BuyOrder: Already processed");
+
+        // disable for gotchi in lending
+        uint256 category = LibAavegotchi.getERC721Category(erc721BuyOrder.erc721TokenAddress, erc721BuyOrder.erc721TokenId);
+        if (category == LibAavegotchi.STATUS_AAVEGOTCHI) {
+            uint32 listingId = s.aavegotchiToListingId[uint32(erc721BuyOrder.erc721TokenId)];
+            require((listingId == 0) || (s.gotchiLendings[listingId].timeAgreed == 0), "ERC721BuyOrder: Not supported for aavegotchi in lending");
+        }
+
+        // hash validation
+        require(
+            erc721BuyOrder.validationHash ==
+                LibBuyOrder.generateValidationHash(erc721BuyOrder.erc721TokenAddress, erc721BuyOrder.erc721TokenId, erc721BuyOrder.validationOptions),
```

Comment:
You could just send through the whole erc721BuyOrder object here, instead of using those 3 params

Reply:
Of course, we can.

Reply:
BTW, do we need to use whole `erc721BuyOrder` object here?
When buy order is created, `validationHash` should be included.

Reply:
Don't quite understand the question @orionstardust

Reply:
@cinnabarhorse 
I think, we don't need to use whole erc721BuyOrder object as a parameter.
Because all validation params are gotchi info, so we need gotchi id and validationOption flags only.